doc/developer/design/20231211_datadog_sink.md

```diff
@@ -0,0 +1,162 @@
+
+#### Timestamp
+
+In Materialize, we have two values to set the timestamp field: `mz_now()` and `now()`. I believe `now()` is enough. The integrations we recommend to our users today use the wall-clock time as a timestamp, so I don’t see any strong reason to switch to `mz_now()` unless we want correctness.
+
+While we could allow the user to set the timestamp, it would make the experience harder. Collections like `mz_cluster_replica_utilization` do not include a timestamp, forcing users to create one themselves. *I do not discard the chance to make it optional.*
+
+### Materialize source structure
+
+A Datadog sink needs a _source_: a table or view — and must have a structure that matches [Datadog's API submit metrics endpoint](https://docs.datadoghq.com/api/latest/metrics/#submit-metrics) fields.
+
+A table in Materialize acting as a source for the sink must have the following structure:
+
+```sql
+CREATE TABLE metrics (
+    -- Required:
+    metric TEXT,
+    resources JSONB,
+    value FLOAT,
+    -- Optional:
+    type INT,
+    interval BIGINT,
+    tags TEXT[],
+    unit TEXT,
+    metadata: JSONB,
+);
+```
+
+And a view needs the same structure:
+```sql
+CREATE MATERIALIZED VIEW metrics AS
+    SELECT
+    -- Required:
+    DISTINCT(
+        jsonb_build_object(
+            'cluster_id', R.cluster_id,
+            'cluster_name', C.name,
+        )
+    ) as resources,
+    'cluster_replica_cpu_usage' as metric,
+    cpu_percent::float as value,
+    -- Optional:
+    0 as type,
+        120 as interval,
+    ARRAY['production', 'Materialize'] as tags,
+    'percent' as unit
+    FROM mz_internal.mz_cluster_replica_utilization U
+    JOIN mz_catalog.mz_cluster_replicas R ON (U.replica_id = R.id)
+    JOIN mz_catalog.mz_clusters C ON (R.cluster_id = C.id);
+```
+
+**NOTE**: Metrics and their values do not need to be unique for each timestamp, but the latest value will suppress the other in Datadog. It would be great if the sink could detect this during creation/execution and emit a warning if there are different values at the same point in time for a set of metrics and resources.
+
+Another distinction is that *resources* equal *labels* in the OpenMetrics definition. I kept *resources* to match Datadog’s API, but I like *labels* as a possible replacement.
+
+### OAuth Process
+
+Datadog recommends an OAuth process for their integrations and creating an API key on behalf of the user if the integration needs to write data into Datadog. The token does not have the scope to write data into Datadog, so creating the API key is necessary.
+
+The OAuth process requires additional logic in the console. The whole process is defined in the following sequence diagram:
+
+<details open>
+<summary>Sequence diagram</summary>
+<img width="2129" alt="Sequence Diagram" src="https://github.com/joacoc/materialize/assets/11491779/28dc14cb-a488-4e8b-8c89-ab69accfd858">
```

Comment:
The console doesn't doesn't have a backend state store, so it's going to struggle to do the OAuth exchange, as it won't have anywhere good to store the `code_verifier` across page loads. Maybeee you could do it through cookies or `localStorage`? Seems a bit unusual though.

The alternative would be to add an HTTP endpoint to `environmentd` that handles the generation of the `code_verifier`, the redirect to the Datadog authorization endpoint, the place that's redirect _back_ from the Datadog authorization endpoint, the `POST` to the Datadog API to get an access token, and the creation of the access token secret. Upsides of that approach: the OAuth authentication would be more tightly integrated with the connection itself. We wouldn't have a separate API key secret that could get deleted separately from the connection. Instead we could have a hidden system secret associated with the connection. We'd probably want to introduce syntax like `AUTHORIZE CONNECTION` that returns the URL to visit in your browser to (re)start the OAuth exchange.